lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp

```diff
@@ -4213,4 +4213,317 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
+                  binder.getLoc(),
+                  rewriter.getType<Torch::ListType>(
+                      rewriter.getType<Torch::FloatType>()),
+                  SmallVector<Value>{countFloat});
+              Value cstDtype = rewriter.create<Torch::ConstantIntOp>(
+                  binder.getLoc(), rewriter.getI64IntegerAttr(
+                                       (int)torch_upstream::ScalarType::Float));
+              SmallVector<int64_t> countShape{1};
+              auto countType = rewriter.getType<Torch::ValueTensorType>(
+                  countShape, resultType.getOptionalDtype());
+              Value countTensor = rewriter.create<Torch::AtenTensorOp>(
+                  binder.getLoc(), countType, dataList, /*dtype=*/cstDtype,
+                  /*layout=*/none, /*requires_grad=*/cstFalse);
+
+              if (is_2d) {
+                // ex: output: [[0,0],[0,0]] ->
+                // outputSequence: [[0,0]] , countUnsqueezed: [[1]] ->
+                // outputSequence:
+                // [[0,1]] -> output: [[0,1],[0,0]]
+                SmallVector<int64_t> outputSequenceShape;
+                outputSequenceShape.push_back(1);
+                outputSequenceShape.push_back(resultShape[1]);
+                auto outputSequenceType =
+                    rewriter.getType<Torch::ValueTensorType>(
+                        outputSequenceShape, resultType.getOptionalDtype());
+                Value batchPlusOne = rewriter.create<Torch::AtenAddIntOp>(
+                    binder.getLoc(), batchValue, one);
+                Value outputSequence =
+                    rewriter.create<Torch::AtenSliceTensorOp>(
+                        binder.getLoc(), outputSequenceType, output,
+                        /*dim=*/zero, batchValue, batchPlusOne, one);
+                Value countUnsqueezed = rewriter.create<Torch::AtenUnsqueezeOp>(
+                    binder.getLoc(),
+                    rewriter.getType<Torch::ValueTensorType>(
+                        llvm::SmallVector<int64_t>{1, 1},
+                        resultType.getDtype()),
+                    countTensor, zero);
+                Value insertStart = rewriter.create<Torch::ConstantIntOp>(
+                    binder.getLoc(),
+                    rewriter.getI64IntegerAttr(ngram_indexes[ngram_i]));
+                Value insertEnd = rewriter.create<Torch::AtenAddIntOp>(
+                    binder.getLoc(), insertStart, one);
+                outputSequence = rewriter.create<Torch::AtenSliceScatterOp>(
+                    binder.getLoc(), outputSequenceType, outputSequence,
+                    countUnsqueezed,
+                    /*dim=*/one, insertStart, insertEnd, /*step=*/one);
+                output = rewriter.create<Torch::AtenSliceScatterOp>(
+                    binder.getLoc(), resultType, output, outputSequence,
+                    /*dim=*/zero, batchValue, batchPlusOne, /*step=*/one);
```

Comment:
When batches exist (2d), then for a given batch, you should only need to insert it once into the output tensor right at the end of the batch loop. There shouldn't be any need to repeatedly slice it from the batched output, update the slice, and rewrite the whole slice back into the batched output tensor for each ngram.

1. At the beginning of the batch loop, set up an initial (rank 1) tensor for that individual batch (call it `outputForBatch` or something). For 2-D, this can be a new rank-1 tensor of zeros (or a slice from the original output init), for 1-D this could be the whole output (`outputForBatch = output`). 
2. in this section where you are writing (slice scattering) to the `output` tensor, instead write to `outputForBatch` tensor. The logic will all match the "else" case here, since `outputForBatch` is rank-1. 
3. At the end of the batch loop, insert the `outputForBatch` slice back into the `output` for 2d, and just set `output=outputForBatch` in 1-D.